src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs

```diff
@@ -71,31 +71,22 @@ protected async Task WaitForHigherPriorityOperationsAsync()
                     {
                         using (Logger.LogBlock(FunctionId.WorkCoordinator_WaitForHigherPriorityOperationsAsync, CancellationToken))
                         {
-                            do
+                            while (true)
                             {
-                                // Host is shutting down
-                                if (CancellationToken.IsCancellationRequested)
-                                {
-                                    return;
-                                }
+                                CancellationToken.ThrowIfCancellationRequested();
 
                                 // we wait for global operation and higher queue operation if there is anything going on
-                                if (!GlobalOperationTask.IsCompleted || !HigherQueueOperationTask.IsCompleted)
-                                {
-                                    await Task.WhenAll(GlobalOperationTask, HigherQueueOperationTask).ConfigureAwait(false);
-                                }
-
-                                // if there are no more work left for higher queue, then it is our time to go ahead
-                                if (!HigherQueueHasWorkItem)
-                                {
+                                await HigherQueueOperationTask.ConfigureAwait(false);
```

Comment:
this method (WaitForHigherPriorityOperationsAsync) is only called at teh start of ExecuteAsync.  So there is no need to do any checking around global operations being in flight.  the top level idle processor loop only calls into this when there's no need to delay work.

Reply:
note: this system is entirely suspect to me @mavasani .  We have all these processor subclasses, and all of them then coordinate with each other to make sure only one is running (and the highest priority one at that).  It's unclear to me why we don' just have a top level processor that just has three queues (or one priority queue) and it just grabs the appropriate next item out of each queue to run.

Reply:
I believe this has been there since the original design. I would be fine if we consolidate these into single processor. One thing that I do notice is that `HighPriorityProcessor` is not a sub-type of `GlobalOperationAwareIdleProcessor`, while the normal and low priority processors are global operation aware. They also seem to have very different implementations, I am not sure if any of these variations were introduced due to any performance investigations. That would be my biggest concern around a big refactoring here. Otherwise, I completely agree that this is a super complicated system, and would really love if it can be simplified, or removed in future.

Reply:
yeah, i don't have a good plan here yet.  It's just... extremely complex and confusing.